include/deal.II/fe/fe_enriched.h

```diff
@@ -0,0 +1,545 @@
+#include <deal.II/base/symmetric_tensor.h>
+#include <deal.II/base/quadrature.h>
+
+#include <vector>
+#include <utility>
+#include <numeric>
+
+DEAL_II_NAMESPACE_OPEN
+
+/**
+ * Implementation of a partition of unity finite element which enriches a standard
+ * finite element with an enrichment function multiplied with another (usually
+ * linear) finite element:
+ * \f[
+ * U(x) = \sum_i N_i(x) U_i + \sum_j N_j(x) \sum_k F_k(x) U_{jk}
+ * \f]
+ * where \f$ N_i(x) \f$ and \f$ N_j(x) \f$ are the underlying finite elements (including
+ * the mapping from the isoparametric element to the real element); \f$ F_k(x) \f$
+ * are the enrichment functions in real space (e.g. \f$ 1/r \f$, \f$ exp(-r) \f$, Heaviside, etc);
+ * \f$ U_i \f$ and \f$ U_{jk} \f$ are the standard and enriched DoFs.
+ *
+ * The simplest constructor of this class takes two finite element objects and an
+ * enrichment function to be used. For example
+ *
+ * @code
+ * FE_Enriched<dim> fe(FE_Q<dim>(1),
+ *                     FE_Q<dim>(1),
+ *                     function)
+ * @endcode
+ *
+ * The \f$ exp(-x) \f$ enrichment leads to the following shape functions on unit
+ * element:
+ * <table>
+ *   <tr>
+ *     <td align="center">
+ *       @image html fe_enriched_1d.png
+ *     </td>
+ *     <td align="center">
+ *       @image html fe_enriched_h-refinement.png
+ *     </td>
+ *   </tr>
+ *   <tr>
+ *     <td align="center">
+ *       1d element, base and partition-of-unity shape functions.
+ *     </td>
+ *     <td align="center">
+ *       enriched shape function resulting from h-refinement of a single element (bottom-left corner).
+ *     </td>
+ *   </tr>
+ * </table>
+ *
+ * <h3>Implementation</h3>
+ *
+ * The implementation of the class is based on FESystem which is aggregated as
+ * a private member and initialized for the case above as
+ *
+ * @code
+ * FESystem<dim> fe_system(FE_Q<dim>(1),1,
+ *                         FE_Q<dim>(1),1);
+ * @endcode
+ *
+ * Note that it would not be wise to have this class derived
+ * from FESystem as the latter concatenates the given elements into different
+ * components of a vector element, whereas the current class combines the given
+ * elements into the same components. For instance, if two scalar elements are
+ * given, the resulting element will be scalar rather than have two components
+ * when doing the same with an FESystem.
+ *
+ * The ordering of the shape function, @p interface_constrains, the @p prolongation (embedding)
+ * and the @p restriction matrices are taken from the FESystem class.
+ *
+ * <h3>hp-constraints</h3>
+ *
+ * This class implements hp constraints with arbitrary FE. For the example above,
+ * a user will need to wrap a given FiniteElement into FE_Enriched:
+ * @code
+ *   FE_Enriched<dim> fe_enriched(FE_Q<dim>(1));
+ * @endcode
+ *
+ * This corresponds to building constraints between enriched FE and
+ *
+ * @code
+ * FESystem<dim>(FE_Q<dim>(1),1,
+ *               FE_Nothing<dim>(1,true),1)
```

Comment:
constraints are borrowed from FESystem. So constrants make sure that each component of two FESystems are continuous. This in turn will make sure that scalar FE space produced by FEEnriched (with and without enrichment) is also continuous.